[WIP] Monitoring tutorial

mikeminutillo opened the pull request:
Getting started content for the new monitoring bits.

- [x] Move snippets
- [ ] Create downloadable sample
- [ ] Take Screenshots
- [ ] Check TODOs

mikeminutillo commented:
@Particular/devops-launch-wave-squad @Particular/docs-maintainers @Particular/editors Please have a quick review to see if I'm on the right track.

Would it make sense to split this up in to multiple parts?

- Component overview
- Installation and configuration
- Using the downloaded sample to explore the UI

adamralph commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+## Component overview
+
+The following diagram shows all of the components that are involved in monitoring an NServiceBus solution.
```
> The following diagram shows all ~of~ the components ~that are~ involved in monitoring an NServiceBus solution.

adamralph commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+![the above renders like this](component-overview.png)
+
+Each endpoint in the solution is configured to collect metrics about it's own operation using the NServiceBus.Metrics NuGet package. This package hooks into the NServiceBus pipeline and calculates key metric data. It also adds headers to outgoing messages to track queue length.
```
> ...and calculates key metric data.

This sentence sounds stunted. It feels like it should have "... using X", "...to Y", or "...which is sent to Z" or something similar appended to it.

mikeminutillo replied:
It's a weird situation because the package doesn't do anything with the data, it just makes it available. I changed it to: 

> This package hooks into the NServiceBus pipeline to calculate and expose key metric data.

adamralph commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+Each endpoint in the solution is configured to collect metrics about it's own operation using the NServiceBus.Metrics NuGet package. This package hooks into the NServiceBus pipeline and calculates key metric data. It also adds headers to outgoing messages to track queue length.
+
+The NServiceBus.Metrics.ServiceControl NuGet package collects the generated metric data and periodically sends it to a centralized monitoring queue. It does this with the same transport that the endpoints use to exchange business messages.
```
> It does this ~with~ _using_ the same transport...

adamralph commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+Each endpoint in the solution is configured to collect metrics about it's own operation using the NServiceBus.Metrics NuGet package. This package hooks into the NServiceBus pipeline and calculates key metric data. It also adds headers to outgoing messages to track queue length.
+
+The NServiceBus.Metrics.ServiceControl NuGet package collects the generated metric data and periodically sends it to a centralized monitoring queue. It does this with the same transport that the endpoints use to exchange business messages.
```
Do we need to explain anywhere what happens if the endpoint goes down during that period? The immediate question I have here is: what happens if the endpoint goes down after data was collected but before it was sent? Does it get sent or lost? Does it matter?

mikeminutillo replied:
It gets lost. I added the phrase "in memory" to make this more explicit

adamralph commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+The NServiceBus.Metrics.ServiceControl NuGet package collects the generated metric data and periodically sends it to a centralized monitoring queue. It does this with the same transport that the endpoints use to exchange business messages.
+
+A ServiceControl Monitoring Instance reads the messages from the monitoring queue and aggregates monitoring data for all of your endpoints. Each environment should have it's own Monitoring Instance. Monitoring Instances are created and managed with the ServiceControl Management Utility. 
```
What is a "ServiceControl Monitoring Instance"? Is that something new? A special instance of ServiceControl? Do we need to preempt this question?

mikeminutillo replied:
It's a new thing. The direction we are heading is that "ServiceControl" is a suite of tools. One of those tools is a Monitoring Instance. We say something to this effect below in the section "Create Monitoring Instance". Here we are describing boxes on a diagram. Do we need to go into this greater depth yet?

adamralph commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+### Install ServicePulse
+
+ServicePulse is a web application for production monitoring and recoverability. To enable it's monitoring features, you must connect ServicePule to a ServiceControl Monitoring Instance.
```
typo: "ServicePule"

tmasternak commented on tutorials/monitoring/Solution/placeholder.txt:
```diff
@@ -0,0 +1,10 @@
+
+Outstanding questions:
+- Should the endpoints already be configured to send data to Particular.Monitoring?
+  - If so, should this configuration be commented out?
```
I would go with a comment in the code saying:
"Put the monitoring plugin configuration here" or sth similar.

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+This tutorial will guide you through getting started with the monitoring capabilities built in to the Particular Service Platform.
+
+You can follow this tutorial to configure any NServiceBus based solution for monitoring with the Particular Service Platform. If you do not have a solution to monitor just yet, you can get started by downloading our sample solution.
```
Should we use `system` instead of `solution`? For me solution sounds pretty VS centric.

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+![the above renders like this](component-overview.png)
+
+Each endpoint in the solution is configured to collect metrics about it's own operation using the NServiceBus.Metrics NuGet package. This package hooks into the NServiceBus pipeline to calculate and expose key metric data. It also adds headers to outgoing messages to track queue length.
```
The `queue-length` headers are added by `NServiceBus.Metrics.ServiceControl`. The idea was that keeping it in `NServiceBus.Metrics` and exposing to the users does not bring much value as a central collection point is requried to calcuate estimates.

adamralph replied:
![image](https://user-images.githubusercontent.com/677704/31216034-4f44351a-a9b2-11e7-82f4-b46dc9419fd3.png)

It could just be the "mysterious" term "key metric data" that is bothering me. I'd probably feel more comfortable if it read something like:

> This package hooks into the NServiceBus pipeline to calculate and expose metrics will can tell us about the health and performance of an endpoint.

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+The NServiceBus.Metrics.ServiceControl NuGet package collects the generated metric data in memory and periodically sends it to a centralized monitoring queue. It does this using the same transport that the endpoints use to exchange business messages.
+
+A ServiceControl Monitoring Instance reads the messages from the monitoring queue and aggregates monitoring data for all of your endpoints. Each environment should have it's own Monitoring Instance. Monitoring Instances are created and managed with the ServiceControl Management Utility. 
```
Should we add links to Monitoring Instance and SCMU?

adamralph replied:
![image](https://user-images.githubusercontent.com/677704/31216288-624ea0a4-a9b3-11e7-9bbe-b4e5ac0a2b12.png)

It could be enough to start this paragraph with a sentence saying something like:

> A ServiceControl Monitoring instance is a new type of ServiceControl instance which....

Bear in mind that there is a high chance that the reader is already familiar with ServiceControl, but it's the first time they have seen the term "ServiceControl Monitoring instance" (it was the first time I'd seen it, hence my question) it feels awkward to use see a new term X without even a short "X is a..." phrase.

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+### Create Monitoring Instance
+
+ServiceControl is a suite of back-end tools that collect useful information about your running system.
```
I really like this sentence. It should probably make it's way to SC doco.

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+ServiceControl is a suite of back-end tools that collect useful information about your running system.
+
+Download and run the latest ServiceControl installer. This installer contains the ServiceControl Management Utility which can be used to create and manage Monitoring Instances. 
```
Do we want to make `Download` a link?

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+SCREENSHOT - ServiceControl Installer
+
+TODO: For the pre-release version it is currently required to add a config file with a feature toggle to the SCMU. We should either document this step or remove it.
```
I don't think we have anything specific in the interation backlog to change that. As a result I would vote to document this step.

mikeminutillo replied:
Given that there is no other config required for SCMU, should we just allow users to download the file? 

I think it would be rather trivial to remove, especially if we add https://github.com/Particular/LaunchWave.DevOpsAdvancedMonitoring/issues/250#issuecomment-333761759 which gives us the opportunity to add an inline "PRE RELEASE" tag to it

tmasternak replied:
So the flow would be:
 * Download SC installer
 * Download `App.config` and put it in SCMU folder
 * After that the `Add instance` screen would show SC.Monitoring with `PRE Release` tag

If I'm getting this right, I'm 👍

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+ServicePulse is a web application for production monitoring and recoverability. To enable it's monitoring features, you must connect ServicePulse to a ServiceControl Monitoring Instance.
+
+Download and run the latest ServicePulse installer. On the ServicePulse Configuration screen check the Monitoring option and enter the location of the monitoring instance API.
```
We need a link and note on how to run with `Monitoring Capability` toogle on. https://github.com/Particular/LaunchWave.DevOpsAdvancedMonitoring/blob/master/alpha-release/installation-guide/index.md#run-the-servicepulse-installer

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+### Configure endpoints
+
+There are two NuGet packages to install into each endpoint. 
```
Should we turn this around and start with `NSB.Metrics.ServiceControl`. We can add note that it depends on `NSB.Metrics` and thats why 2 packages are installed but I'm not sure that going into details about `NSB.Metrics` is important from the tutorial perspective - btw we have that explained in the diagram as well.

mikeminutillo replied:
Good point. It lets us collapse the snippet as well.

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+There are two NuGet packages to install into each endpoint. 
+
+TODO: Should the endpoints in the sample solution for this tutorial be already configured to send monitoring data OR should we explain how to do it?
```
I would explain how to do it and leave comments in code as mentioned in https://github.com/Particular/docs.particular.net/pull/3299/files#r142863261

tmasternak commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+Once the Management Utility is installed, run it.
+
+TODO: For the pre-release version it is currently required to install a valid license. We should either document this step or remove it.
```
We do have a timebomb instead of licence item in current iteration to I would add a not on the 3 months validity of Monitoring Instance.

adamralph commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+---
+title: Monitoring NServiceBus solutions
+reviewed: 2017-10-02
+summary: Getting started with the monitoring capabilities built in to the Particular Service.
```
Missing "Platform".

adamralph commented on tutorials/monitoring/tutorial.md:
```diff
@@ -0,0 +1,141 @@
+
+Each endpoint in the solution is configured to collect metrics about it's own operation using the NServiceBus.Metrics NuGet package. This package hooks into the NServiceBus pipeline to calculate and expose key metric data. It also adds headers to outgoing messages to track queue length.
+
+The NServiceBus.Metrics.ServiceControl NuGet package collects the generated metric data in memory and periodically sends it to a centralized monitoring queue. It does this using the same transport that the endpoints use to exchange business messages.
```
![image](https://user-images.githubusercontent.com/677704/31216197-0dbd001c-a9b3-11e7-8fb8-1d373482b407.png)

While that makes it more explicit, as a reader I may still be asking questions like:

> What if the endpoint goes down and loses the data before it is sent? Sounds a bit sketchy to me...

Do we need to preempt those types of questions anywhere?

mikeminutillo commented:
Split the existing tutorial content up into 3 sections and added an overview page. I also added the instructions for installing and setting up a ServiceControl instance (which is also required at this point).

mikeminutillo commented:
Woo!

weralabaj commented:
@mikeminutillo go to sleep!

tmasternak commented on tutorials/monitoring/1-component-overview/index.md:
```diff
@@ -0,0 +1,69 @@
+
+Each system being monitored will push data to three main queues:
+
+- Monitoring: Each endpoint will calculate and send metric data to this queue. This data can tell us about the health and performance of an endpoint.
```
"gather" instead of "calculate"
"this data holds information about" instead of "this data can tell us"

tmasternak commented on tutorials/monitoring/1-component-overview/index.md:
```diff
@@ -0,0 +1,69 @@
+
+The core NServiceBus package can be configured to send messages to the Audit and Error queues. In order to send data to the Monitoring queue, two additional packages are required:
+
+- NServiceBus.Metrics: hooks into the NServiceBus pipeline to calculate and expose metrics.
```
remove "calculate"

tmasternak commented on tutorials/monitoring/1-component-overview/index.md:
```diff
@@ -0,0 +1,69 @@
+The core NServiceBus package can be configured to send messages to the Audit and Error queues. In order to send data to the Monitoring queue, two additional packages are required:
+
+- NServiceBus.Metrics: hooks into the NServiceBus pipeline to calculate and expose metrics.
+- NServiceBus.Metrics.ServiceControl: collects the metric data generated by NServiceBus.Metrics and periodically sends it to the monitoring queue. This package also instruments outgoing messages with additional headers that are used to estimate queue length.
```
How about we change "collects" to "buffers and serializes metric data ..."

tmasternak commented on tutorials/monitoring/1-component-overview/index.md:
```diff
@@ -0,0 +1,69 @@
+- NServiceBus.Metrics: hooks into the NServiceBus pipeline to calculate and expose metrics.
+- NServiceBus.Metrics.ServiceControl: collects the metric data generated by NServiceBus.Metrics and periodically sends it to the monitoring queue. This package also instruments outgoing messages with additional headers that are used to estimate queue length.
+
+ServiceControl is a suite of back-end tools that collect useful information about your running system. A monitoring solution will contain two different instances:
```
>A monitoring solution will contain two different instances:

how about we write

> A monitoring system back-end consists of two components

tmasternak commented on tutorials/monitoring/1-component-overview/index.md:
```diff
@@ -0,0 +1,69 @@
+ServiceControl is a suite of back-end tools that collect useful information about your running system. A monitoring solution will contain two different instances:
+
+- ServiceControl instance: collects messages from the audit and error queues. Error messages are grouped together and can be retried in bulk.
+- Monitoring instance: a new type of ServiceControl instance that collects data from the monitoring queue about all of the endpoints in an NServiceBus system. 
```
I would skip "a new type of ServiceControl instance that" -

tmasternak commented on tutorials/monitoring/1-component-overview/index.md:
```diff
@@ -0,0 +1,69 @@
+- ServiceControl instance: collects messages from the audit and error queues. Error messages are grouped together and can be retried in bulk.
+- Monitoring instance: a new type of ServiceControl instance that collects data from the monitoring queue about all of the endpoints in an NServiceBus system. 
+
+Both of these instance types can be created and managed with the ServiceControl Management Utility.
```
I would change "created" to "set-up"

tmasternak commented on tutorials/monitoring/2-setting-up-environment/index.md:
```diff
@@ -0,0 +1,100 @@
+
+### Install ServiceControl
+
+ServiceControl is a suite of back-end tools that collect useful information about your running system. The main installer includes a desktop utility called the ServiceControl Managemeny Utility which can be used to create an manage ServiceControl and Monitoring instances.
```
typo in "Managemeny"

tmasternak replied:
"to create an manage" should be "to create and manage"

tmasternak commented on tutorials/monitoring/2-setting-up-environment/index.md:
```diff
@@ -0,0 +1,100 @@
+A ServiceControl instance collects messages from the error queue and manages retrying them. It also collects messages from the audit queue. 
+
+In the ServiceControl Management Utility, click **+ NEW INSTANCE** and select **ServiceControl instance** from the list of options.
+
```
I would add a NOTE here that guides users that have already run out of Trial License (Standard License users). I would emphasize that getting license is easy :). See: https://github.com/Particular/LaunchWave.DevOpsAdvancedMonitoring/issues/266

tmasternak commented on tutorials/monitoring/3-configuring-endpoints/index.md:
```diff
@@ -0,0 +1,85 @@
+
+## Configure endpoints
+
+In order to be monitored by the Particular Service Platform, each endpoint needs to be configured to send data to three different queues: error, audit, and monitoring. It is recommended to go through this entire process for a single endpoint and ensure that it is working before continuing on to the next endpoint.
```
I would put the last sentence in a separate NOTE - I think it's quite important.

tmasternak commented on tutorials/monitoring/2-setting-up-environment/index.md:
```diff
@@ -0,0 +1,100 @@
+
+The listing for the Monitoring instance includes a URL link. This URL will be needed when installing ServicePulse below.
+
+NOTE: Creating the Monitoring instance will also create the monitoring queue if it did not already exist. By default this queue is called Particular.Monitoring. 
```
Particular.Monitoring should have quotes.

tmasternak commented on tutorials/monitoring/3-configuring-endpoints/index.md:
````diff
@@ -0,0 +1,85 @@
+
+You can configure the location of the error queue. Add the following code to your endpoint configuration:
+
+```csharp
````
should this be changed to a snippet ref?

mikeminutillo replied:
Yeah, it will be. I was running into trouble doing that so I in-lined a few of the snippets make the process smooth for now

tmasternak commented on tutorials/monitoring/3-configuring-endpoints/index.md:
````diff
@@ -0,0 +1,85 @@
+
+You can enable auditing and configure the location of the audit queue. Add the following code to your endpoint configuration:
+
+```csharp
````
should this be changed to a snippet ref?

tmasternak commented on tutorials/monitoring/3-configuring-endpoints/index.md:
````diff
@@ -0,0 +1,85 @@
+
+You can then enable monitoring and configure the location of the monitoring queue. Add the following code to your endpoint configuration:
+
+```csharp
````
should this be changed to a snippet ref?

tmasternak commented on tutorials/monitoring/3-configuring-endpoints/index.md:
```diff
@@ -0,0 +1,85 @@
+This third lesson guides you through setting up all of the endpoints in your system to be monitored by the Particular Software Platform.
+
+
+## Configure endpoints
```
How about we state that one is using the sample solution e.g. Billing is a good place to start and add comments in `Program.cs` where say: "this is the place where you should put x-y-z"

tmasternak commented on tutorials/monitoring/3-configuring-endpoints/index.md:
```diff
@@ -0,0 +1,85 @@
+
+metrics.SendMetricDataToServiceControl(
+    serviceControlMetricsAddress: "Particular.Monitoring", 
+    interval: TimeSpan.FromSeconds(10)
```
I would add a comment around the interval stating that it influences only the queue length data reports.

tmasternak commented on tutorials/monitoring/4-endpoint-performance/index.md:
```diff
@@ -0,0 +1,67 @@
+
+Throughput is a measure of how much work the endpoint is doing. It is the rate at which the endpoint is able to process messages from it's input queue. While throughput can reflect endpoint performance, it is also heavily influenced by how much work there is to do. A highly optimized endpoint with only a few messages per second to process will still have a low throughput. 
+
+Processing time is a measure of the time it takes for the endpoint to process a single message. A higher processing time indicates a slower endpoint and a lower processing time indicates a faster endpoint. The amount of work assigned to an endpoint has less of an impact on processing time than it does on throughput. This makes it a better guage for individual endpoint performance.
```
"Processing time is a measure of the time " -> "Processing time is the time ..."

tmasternak commented on tutorials/monitoring/4-endpoint-performance/index.md:
```diff
@@ -0,0 +1,67 @@
+
+include: monitoring-intro-paragraph
+
+This fourth lesson guides you through the throughput and processing time metrics and teaches you how to use those metrics to interpret the performance of an individual endpoint. Examples are shown using the sample monitoring solution.
```
I think we should add an information that we assume that users have configured monitoring in the sample solution as described in lesson 3.

adamralph commented:
I haven't been able to keep up with the latest changes, but if you still want another editorial review once things have stabilised, please feel free to request a review from `@editors` once again.

mikeminutillo commented:
Thanks @tmasternak 

I'm thinking that the solution downloaded from Lesson 0 should contain placeholder comments and we should copy that solution into Lesson 3 with all of the placeholders filled in.

That way, if you are reading lesson 3 you have the option of doing it manually with Solution0 OR bypassing that and downloading the working Solution3. A BEFORE and AFTER type deal.

Then Lessons 4, 5, & 6 can just refer to Solution3.

DavidBoike commented on tutorials/monitoring/2-setting-up-environment/index.md:
```diff
@@ -0,0 +1,102 @@
+
+NOTE: Creating and managing ServiceControl instances requires a license file. If you do not already have a license then the ServiceControl Management Utility will generate a trial license for you. If you have already had a trial license and it has expired, then follow the on-screen prompts to extend your trial license. It is a short form and your new license file will get sent to your email addre ...
+
+Under TRANSPORT CONFIGURATION select the transport that your NServiceBus system runs on. Depending on the transport you may be required to add an additional connection string.
```
FYI my convention with the tutorials thus far (stolen from Packt writing guidelines) is to have all on-screen, non-code text (things you're selecting, clicking, looking for, etc.) all be in bold. So I think that would cover **TRANSPORT CONFIGURATION** here (assume you're using caps because SCMU does?), **SQLServer** in the NOTE below, **AUDIT FORWARDING** + **On** + **Off** in the next paragraph, etc.

It's personally jarring to me to have the caps there. Honestly CONFIGURATION looks misspelled to me all in caps! I had to check it four times. I'd consider doing **Word Case in Bold** in those instances even though SCMU is all-caps.

DavidBoike commented:
I see you're having some issues because there's one solution (and source for snippets?) but multiple logical tutorials that draw from it.

Note that currently, you have the !!tutorial metadata on `index.md` files, which will be treated as articles, so they won't look at or understand that extension metadata and (I believe) it will get ignored.

I'd totally be open to talking about DocsEngine modifications that would make this easier.

What about metadata like

```
- !!tutorial
  solution: root-relative/path-to/solution-directory
  snippets: root-relative/path-to/snippets-directory
```

You could use either, both, or neither. It would just replace the conventions-based directory the LessonRenderer is looking for. Then all of these could be tutorials (tutorial.md) and all of them could point to the same download, or use the same set of snippets.

Snippets might be a bit harder because of the code that checks to see if there are unused snippets, but it would certainly be useful for the Intro tutorial right now, where each of the 5 lessons has to have its own Snippets solution, and it would be a lot more maintainable to have one big Snippets solution in `/tutorials/intro-to-nservicebus/Snippets`.

tmasternak commented:
>  A BEFORE and AFTER type deal.

@mikeminutillo I like this idea.

mikeminutillo commented:
Moved to #3359
